### ISSUE TITLE
Attempting to use containerd-snapshotter mode with nix-snapshotter

### ISSUE THREAD

**Issue by hinshun** [moby/moby#47444]
### Description

I was trying to use [nix-snapshotter](https://github.com/pdtpartners/nix-snapshotter) with Docker but ran into a few issues. We have it working no problem with containerd + rootless containerd so I believe the issues is with Docker's containerd integration.

```sh
[root@nixos:~]# docker run --rm ghcr.io/pdtpartners/hello
docker: Error response from daemon: failed to mount : mkdir /var/lib/docker/rootfs/nix/07b044aabc1d3fb1da00a927e2f269bae39b1e04366434f88bc408331261558b: no such file or directory.
See 'docker run --help'.
```

Manually creating the directory gets me a bit further but then I run into chmod error:

```sh
[root@nixos:~]# mkdir -p /var/lib/docker/rootfs/nix
[root@nixos:~]# docker run --rm ghcr.io/pdtpartners/hello
docker: Error response from daemon: failed to mount : chmod /var/lib/docker/rootfs/nix/b3f74ea450ba2ef592208478828bae37d18dfbb7822a0869e12ce65d1f3279ba: read-only file system.
```

Perhaps related to this part of the code:
https://github.com/moby/moby/blob/v25.0.3/daemon/snapshotter/mount.go#L123-L131

As an aside, I've contributed this PR to containerd back in Dec 2022, which added the `Target` field to containerd's `mount.Mount` struct:
https://github.com/containerd/containerd/pull/7840

I needed it to construct bind mounts to subdirectorys of the target mountpath. Does the `refCountMounter` still make sense in this case?

```go
	if count := m.rc.Increment(target); count > 1 {
		return target, nil
	}
```

TLDR around nix-snapshotter mounts is its an overlayfs on `/` + many bind mounts under `/nix/store/<package-hash>`, see the [architecture docs](https://github.com/pdtpartners/nix-snapshotter/blob/main/docs/architecture.md) for more details.

### Reproduce

I've prepared a docker image that runs QEMU VM with NixOS pre-configured with this reproduction (docker + containerd + nix-snapshotter):

```sh
docker run --rm -it ghcr.io/pdtpartners/nix-snapshotter:docker-24.0.5

nixos login: root # (Ctrl-a then x to quit)
Password: root

[root@nixos:~]# docker image ls
REPOSITORY                        TAG       IMAGE ID       CREATED         SIZE
ghcr.io/pdtpartners/hello         latest    2971ba7e1fad   4 minutes ago   1.78kB
ghcr.io/pdtpartners/redis-shell   latest    6c6c48068af3   4 minutes ago   20.5kB
ghcr.io/pdtpartners/redis         latest    fac949281ac3   4 minutes ago   14.5kB

[root@nixos:~]# docker run --rm ghcr.io/pdtpartners/hello
```

Docker, containerd, and nix-snapshotter are all in their respective systemd units:

```sh
systemctl status docker
systemctl status containerd
systemctl status nix-snapshotter
```

### Expected behavior

Docker run should work with snapshotters that work with containerd.

### docker version

```bash
Client:
 Version:           24.0.5
 API version:       1.43
 Go version:        go1.21.4
 Git commit:        v24.0.5
 Built:             Thu Jan  1 00:00:00 1970
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          24.0.5
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.21.4
  Git commit:       v24.0.5
  Built:            Tue Jan  1 00:00:00 1980
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.7.11
  GitCommit:        v1.7.11
 runc:
  Version:          1.1.8
  GitCommit:
 docker-init:
  Version:          0.19.0
  GitCommit:
```

### docker info

```bash
Client:
 Version:    24.0.5
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.2
    Path:     /nix/store/gdyxcsv9w1aghvgjy1j4c4mmdh35hi93-docker-plugins/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  2.23.1
    Path:     /nix/store/gdyxcsv9w1aghvgjy1j4c4mmdh35hi93-docker-plugins/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 3
 Server Version: 24.0.5
 Storage Driver: nix
  driver-type: io.containerd.snapshotter.v1
 Logging Driver: journald
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: v1.7.11
 runc version:
 init version:
 Security Options:
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.1.69
 Operating System: NixOS 23.11 (Tapir)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 3.839GiB
 Name: nixos
 ID: 724cf269-930a-4bc5-9da5-f93c46987016
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: true
```

### Additional Info

```sh
# daemon.json
{
  "containerd": "/run/containerd/containerd.sock",
  "features": {
    "containerd-snapshotter": true
  },
  "group": "docker",
  "hosts": [
    "fd://"
  ],
  "live-restore": true,
  "log-driver": "journald",
  "storage-driver": "nix"
}
```

```
# containerd.toml
version = 2
[plugins]
[plugins."io.containerd.grpc.v1.cri"]
[plugins."io.containerd.grpc.v1.cri".cni]
bin_dir = "/nix/store/afw4hbr9fy6983hph8q5isrwa6pwfrrq-cni-plugins-1.3.0/bin"

[plugins."io.containerd.grpc.v1.cri".containerd]
snapshotter = "nix"

[[plugins."io.containerd.transfer.v1.local".unpack_config]]
platform = "linux/amd64"
snapshotter = "nix"

[proxy_plugins.nix]
address = "/run/nix-snapshotter/nix-snapshotter.sock"
type = "snapshot"
```

**Comment by vvoland** [comment]
Thanks, that's a really good and detailed report!

Yeah this code was mostly based on the previous graphdriver implementation and it assumed that these mounts are writable.

I'll need some more time to dig into it deeper 👀

**Comment by hinshun** [comment]
@vvoland Is there anything I can help with to move this along? I think I still need guidance on the moby counterparts, happy to discuss in more detail if the problem is not clear.

**Comment by vvoland** [comment]
The refcounted mounter makes sure that we don't mount the same container rootfs more than once, so it still makes sense in this case.

There were some changes to the `daemon/snapshotter/mount.go` file since v24 - mainly using the refcounted mounter by default - in your case I think the non-refcounted mounter was used, which [did a plain `mkdir` ](https://github.com/moby/moby/blob/v24.0.5/daemon/snapshotter/mount.go#L131) which failed and didn't finish the whole code execution.

Starting from v25 this non-mkdir-all path is gone and the mkdir should succeed.

Unfortunately I haven't had the time yet to dig more deeply into this, but looking briefly at the current master code, I don't see an immediate issue.

Could you please try the `docker run` part with v26?
